### DISCUSSION TITLE
No name for properties in struct return data after calling contract

### DISCUSSION THREAD

**Discussion by aurelien-cuvelier** [wevm/viem#1656] [Idea / Feature Request]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

1.19.0

### Current Behavior

When calling a contract function which will return something more complex than a single value (basically a struct), the returned data won't be formatted correctly with the name of the returned properties.

When calling "slot0" on an Uniswap V3 pool:

```
const poolContract = getContract({
        abi: V3POOL_ABI,
        address: 'UNISWAP_V3_POOL_ADDR',
        publicClient: client,
      });
      
const slot0_res = await poolContract.read.slot0();
```

I get the following array prototype:
```
0:60315243730365501172258484916349n
1:132706
2:0
3:1
4:1
5:0
6:true
```

The result is the same when i use readContract directly from the client, or even after decoding the raw data from a client.call, using "decodeFunctionResult".

### Expected Behavior

The expected behavior would have been to actually be able to call props in the return data by their name, just like ethersjs does:

```
sqrtPriceX96:60315243730365501172258484916349n
tick:132706
observationIndex:0
observationCardinality:1
observationCardinalityNext:1
feeProtocol:0
unlocked:true
```

### Steps To Reproduce

Calling any contract's function which will return a struct.

**Comment by NBMSacha** [comment]
+++ 

Really annoying

**Comment by xyeuu** [comment]
omg insane...

**Comment by jxom** [comment]
Please see https://viem.sh/docs/faq.html#why-is-a-contract-function-return-type-returning-an-array-instead-of-an-object

**Comment by jxom** [comment]
Though, you did mention that the function returns a struct. In which, it should decode as an object. Got an ABI that I can look at?

**Comment by anouci** [comment]
Same

**Comment by jxom** [comment]
Looks like you may be running into this nuance with ABI compilation: https://x.com/_jxom/status/1645391096204369920

Maybe we could add a `returnAs: “object”` API or something for these cases.

**Comment by aurelien-cuvelier** [comment]
> Though, you did mention that the function returns a struct. In which, it should decode as an object. Got an ABI that I can look at?

Sure, here is the ABI for the slot0:

```
  {
    inputs: [],
    name: 'slot0',
    outputs: [
      {internalType: 'uint160', name: 'sqrtPriceX96', type: 'uint160'},
      {internalType: 'int24', name: 'tick', type: 'int24'},
      {internalType: 'uint16', name: 'observationIndex', type: 'uint16'},
      {internalType: 'uint16', name: 'observationCardinality', type: 'uint16'},
      {
        internalType: 'uint16',
        name: 'observationCardinalityNext',
        type: 'uint16',
      },
      {internalType: 'uint8', name: 'feeProtocol', type: 'uint8'},
      {internalType: 'bool', name: 'unlocked', type: 'bool'},
    ],
    stateMutability: 'view',
    type: 'function',
  }
```
